src/core/net/srp_client.hpp

```diff
@@ -0,0 +1,731 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef SRP_CLIENT_HPP_
+#define SRP_CLIENT_HPP_
+
+#include "openthread-core-config.h"
+
+#include <openthread/srp_client.h>
+
+#include "common/clearable.hpp"
+#include "common/linked_list.hpp"
+#include "common/locator.hpp"
+#include "common/message.hpp"
+#include "common/non_copyable.hpp"
+#include "common/notifier.hpp"
+#include "common/timer.hpp"
+#include "crypto/ecdsa.hpp"
+#include "net/dns_headers.hpp"
+#include "net/ip6.hpp"
+#include "net/udp6.hpp"
+
+#if OPENTHREAD_CONFIG_SRP_CLIENT_ENABLE
+
+/**
+ * @file
+ *   This file includes definitions for the SRP (Service Registration Protocol) client.
+ */
+
+namespace ot {
+namespace Srp {
```

Comment:
I had this but removed it intentioanlly.
- I think `Srp` itself is good/clear namespace for it (`Srp::Client` reads better than `Dns::Srp::Client`)
- It allows to re-use some method/type names without worrying about conflicts.
- I think explicitly using `Dns::{Type}` when using them in SRP module would help with it readability, for example I prefer `Dns::Name::ParseName()` instead of `Name::ParseName()` in the SRP code.

Thoughts?

Reply:
It does make some sense to me. Fine to keep it as is for now. 

Still here are some thoughts, just for discussion

As we know, SRP uses the standard DNS update standard mechanism to enable DNS-based Service Discovery using unicast - It easily make people think it as some kind of DNS update implementation, as well from implementation's perspective, it should be able to use the common data structure like `class Name`, thus it seems more reasonable to be under the bigger `Dns` namespace scope. 

Instead of the nested namespaces suggested above, another option we may consider is a flat `Dns` namespace with renamed `DnsClient` and `SrpClient `class, which naturally uses the common type/methods like `Name::ParseName()` without explicitly specifying the namespace `Dns`.

Thoughts?

Reply:
I personally prefer `Dns::Clinet` over `DnsClinet`. It also keeps it consistent with other protocols `Sntp::Clinet`, `Dhcp6::Clinet`, etc.